napi_basics.md

```diff
@@ -0,0 +1,120 @@
+# N-API - Basics
+
+N-API is an API exposed by Node to developers that allow them to author Native add-ons 
+in an ABI stable manner. This means that these add-ons can work against future versions
+of Node without recompilation, and developers get the additional benefit that their 
+add-ons can work independent of the underlying JavaScript Virtual Machine being used 
+by Node.
+
+APIs exposed in the N-API surface are generally used to create and manipulate JavaScript
+objects. Concepts and operations generally map to ideas specified in the ECMA262 Language
+SPecification. The APIs generally have the following properties:
+- All N-API APIs generally return a status code of type `napi_status`. This status 
+can be used to determine whether the API call succeeded or failed.
+- The actual value being returned by the API is returned in an out parameter.
+- All JavaScript objects are abstracted behind an opaque type named `napi_value`.
+- In case of an error status code, additional information can be obtained using 
+`napi_get_last_error_info`
+
+## N-API types
+
+N-API exposes the following fundamental datatypes as abstractions that are consumed by 
+the various APIs. These APIs should be treated as opaque types that can be introspected
+only with other N-API APIs.
+
+### *napi_env*
+
+`napi_env` is used to represent a context that the underlying N-API implementation can use 
+to persist VM specific state.
+<< REVIEW: What else can we say here? >>
+
+### *napi_value*
+This is an opaque pointer that is used to represent a JavaScript object.
+
+### *napi_ref*
+This is the abstraction to use to reference a `napi_value`. This allows for users 
+to manage the lifetimes of JavaScript objects, including defining their minimum
+lifetimes explicitly, and also being notified when they are garbage-collected.
```

Comment:
All the APIs that wrap some kind of external data support finalize callbacks: `napi_wrap()`, `napi_create_external()`, `napi_create_external_buffer()`, `napi_create_external_arraybuffer()`. But the finalize callbacks are not necessarily associated with a reference, and it is possible to create a reference to an ordinary JavaScript value where there is no ability to have a finalize callback. (V8 does support weak-ref callbacks for any kind of JS values, but that is not exposed in N-API.)

I'd suggest removing the mention of the finalize callback from this section about `napi_ref`, and adding more detail to the section about the `napi_finalize` callback, describing how the callback is used specifically to clean up external data that is wrapped by JS values.